File: ios/Sources/Web2AppSDK/AppCallbackProducer.swift
import Foundation

#if canImport(UIKit)
import UIKit
#endif

/// APP_INSTALLED-продюсер (закрывает Lucas-дыру «висячий хук»). На 1-м запуске любая
/// ветка резолва guid → `POST /public/handoff/app-callback` (метрика conversionToApp).
/// Идемпотентно на бэке (204). Grant НЕ зависит от этого callback — чисто метрика.
struct AppCallbackProducer {
    let config: Web2AppConfig

    func reportAppInstalled(guid: String) {
        let url = config.baseUrl.appendingPathComponent("public/handoff/app-callback")
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")

        // Тело зеркалит AppCallbackDto: { guid, projectId, device, event }.
        let body: [String: String] = [
            "guid": guid,
            "projectId": config.projectId,
            "device": Self.platform,
            "event": "app_installed",
        ]
        req.httpBody = try? JSONSerialization.data(withJSONObject: body)

        // Fire-and-forget: сбой метрики НЕ ломает пользовательский поток.
        URLSession.shared.dataTask(with: req).resume()
    }

    /// device-строка для дедуп-ключа метрики (bounded whitelist на бэке: ios|android).
    static var platform: String { "ios" }
}

File: android/src/main/kotlin/app/web2app/sdk/AppCallbackProducer.kt
package app.web2app.sdk

import org.json.JSONObject

/**
 * APP_INSTALLED-продюсер (закрывает Lucas-дыру «висячий хук»). На 1-м запуске любая
 * ветка резолва guid → `POST /public/handoff/app-callback` (метрика conversionToApp, 204,
 * идемпотентно на бэке). Grant НЕ зависит от этого callback — чисто метрика, fire-and-forget.
 */
internal class AppCallbackProducer(private val config: Web2AppConfig) {
    fun reportAppInstalled(guid: String) {
        Http.io {
            val body = JSONObject()
                .put("guid", guid)
                .put("projectId", config.projectId)
                .put("device", "android")
                .put("event", "app_installed")
                .toString()
            // Сбой метрики НЕ ломает пользовательский поток.
            Http.postJson("${config.baseUrl}/public/handoff/app-callback", body)
        }
    }
}
